civis/io/_tables.py

```diff
@@ -1335,3 +1336,58 @@ def _process_cleaning_results(cleaning_futures, client, headers,
     if need_table_columns:
         table_columns = _replace_null_column_names(table_columns)
     return cleaned_file_ids, headers, compression, delimiter, table_columns
+
+
+def _check_column_types(table_columns, file_columns, output_obj_id):
+    """Check that base column types match those current defined for the table.
+
+    Parameters
+    ----------
+
+    table_columns: List[Dict[str, str]]
+      The columns for the table to be created.
+    file_columns: List[Dict[str, str]]
+      The columns detected by the Civis API for the file.
+    output_obj_id: int
+      The file ID under consideration; used for error messaging.
+
+    Raises
+    ------
+
+    CivisImportError
+      If the table columns and the file columns have a type mismatch, or
+      differ in count.
+    """
+    if len(table_columns) != len(file_columns):
+        raise CivisImportError('All files should have the same number of '
+                               'columns. Expected {} columns but file {} '
+                               'has {} columns'.format(
+                                   len(table_columns),
+                                   output_obj_id,
+                                   len(file_columns))
+                               )
+
+    error_msgs = []
+    for idx, (tcol, fcol) in enumerate(zip(table_columns, file_columns)):
+        # for the purposes of type checking, we care only that the types
+        # share a base type (e.g. INT, VARCHAR, DECIMAl) rather than that
```

Comment:
It looks like this check will cause an error for different integer types. Would you want this to fail if one is `INTEGER` and one is `BIGINT`? Also, is there any possibility you'd get different aliases passed in here? e.g. one is `INT` and one is `INT4`, and it throws an error?

Reply:
Something else that occurred to me... is this interoperable with Postgres?

Reply:
1. We wanted to keep our scope as limited here as possible, which means that we are not processing types behind those with explicit lengths like `VARCHAR`. We are accepting that the type comparisons are not going to be perfect as the price we pay for avoiding rewriting the database's type inheritance system. The easiest way to get around the issue of type inference is to pre-create a table with the types that a user knows are correct. 

As far as your comment about aliases like `INT` vs `INT4`, the underlying system that performs the inference Platform-side will only return a single type from a given alias family, so that is not a concern.

2. This should be interoperable with Postgres - the preprocessing job is unaware of any types specific to a database. The types returned here are generic across the databases we support.